sources/httpUtils.ts

```diff
@@ -30,60 +26,55 @@ export async function fetchUrlStream(url: string, options: RequestOptions = {})
     }
   }
 
-  return new Promise<IncomingMessage>((resolve, reject) => {
-    const createRequest = (url: string) => {
-      const request: ClientRequest = https.get(url, {...options, agent: proxyAgent}, response => {
-        const statusCode = response.statusCode;
-
-        if ([301, 302, 307, 308].includes(statusCode as number) && response.headers.location)
-          return createRequest(response.headers.location as string);
-
-        if (statusCode != null && statusCode >= 200 && statusCode < 300)
-          return resolve(response);
-
-        return reject(new Error(`Server answered with HTTP ${statusCode} when performing the request to ${url}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`));
-      });
+  let response;
+  try {
+    response = await globalThis.fetch(input, {
+      ...init,
+      dispatcher: agent,
+    });
+  } catch (error) {
+    throw new Error(
+      `Error when performing the request to ${input}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`,
+      {cause: error},
+    );
+  }
 
-      request.on(`error`, err => {
-        reject(new Error(`Error when performing the request to ${url}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`));
-      });
-    };
+  if (!response.ok) {
+    await response.arrayBuffer();
+    throw new Error(
+      `Server answered with HTTP ${response.status} when performing the request to ${input}; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting`,
+    );
+  }
 
-    createRequest(url);
-  });
+  return response;
 }
 
-export async function fetchAsBuffer(url: string, options?: RequestOptions) {
-  const response = await fetchUrlStream(url, options);
-
-  return new Promise<Buffer>((resolve, reject) => {
-    const chunks: Array<Buffer> = [];
+export async function fetchAsJson(input: string | URL, init?: RequestInit) {
+  const response = await fetch(input, init);
+  const json: any = await response.json();
+  return json;
+}
 
-    response.on(`data`, chunk => {
-      chunks.push(chunk);
-    });
+export async function fetchUrlStream(input: string | URL, init?: RequestInit) {
+  const response = await fetch(input, init);
+  const webStream = response.body;
+  assert(webStream, `Expected stream to be set`);
+  const stream = Readable.fromWeb(webStream);
+  return stream;
+}
 
-    response.on(`error`, error => {
-      reject(error);
-    });
+async function getProxyAgent(input: string | URL) {
+  const {getProxyForUrl} = await import(`proxy-from-env`);
 
-    response.on(`end`, () => {
-      resolve(Buffer.concat(chunks));
-    });
-  });
-}
+  const proxy = getProxyForUrl(input.toString());
```

Comment:
nit: in case `input` is not something we expect and doesn't have a `toString` method, you'd get `toString is not a function`. Instead, you can get a more precise `TypeError: Cannot convert object to primitive value` using a string template.
```suggestion
  const proxy = getProxyForUrl(`${input}`);
```

Reply:
Using a string template allows other cases we don't want though:

```console
$ node
Welcome to Node.js v21.6.1.
Type ".help" for more information.
> (undefined).toString()
Uncaught TypeError: Cannot read properties of undefined (reading 'toString')
> `${undefined}`
'undefined'
> (null).toString()
Uncaught TypeError: Cannot read properties of null (reading 'toString')
> `${null}`
'null'
>

```

Reply:
Fair enough, there's always a tradeoff, pick your poison I guess. FWIW, in Node.js core, we always use string template so the code is not affected is `String.prototype.toString` was mutated. Some V8 versions also was able to better optimize the string to string conversion, not sure if it still applies.